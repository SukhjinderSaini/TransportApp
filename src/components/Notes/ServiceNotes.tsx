import React, { useState, useEffect, useMemo } from "react";
import { Form, Button } from "react-bootstrap";
import ListGroup from "react-bootstrap/ListGroup";

interface Props {
  serviceName: string;
}

function ServiceNotes({ serviceName }: Props) {
  const [notes, setNotes] = useState<string[]>([]);
  const [newNote, setNewNote] = useState("");

  const validateAndAdd = () => {
    if (newNote && newNote.length > 0) {
      const note = newNote;
      addNote(note);
    }
    setNewNote("");
  };

  const textAreaChange = (event: { target: { value: string } }) => {
    const value = event.target.value;
    setNewNote(value);
  };

  const storedNotes = useMemo(() => {
    return localStorage.getItem(serviceName + "_notes")
      ? JSON.parse(localStorage.getItem(serviceName + "_notes"))
      : [];
  }, [serviceName]);

  const addNote = (notesValue: string) => {
    let existingNotes = localStorage.getItem(serviceName + "_notes")
      ? JSON.parse(localStorage.getItem(serviceName + "_notes"))
      : [];
    existingNotes = [...existingNotes, notesValue];
    setNotes(existingNotes);
    localStorage.setItem(serviceName + "_notes", JSON.stringify(existingNotes));
  };

  useEffect(() => {
    setNotes(storedNotes);
  }, [storedNotes]);

  return (
    <>
      {notes.length > 0 && (
        <div className="mt-3 text-start">
          <h5>Notes for {serviceName}</h5>
        </div>
      )}
      <ListGroup>
        {notes.length > 0 &&
          notes.map((note, index) => (
            <ListGroup.Item className="text-start text-break" key={index}>
              <span>{note}</span>
            </ListGroup.Item>
          ))}
      </ListGroup>
      <Form.Control
        value={newNote}
        onChange={textAreaChange}
        className="mt-3 mb-3"
        as="textarea"
        aria-label="With textarea"
        placeholder="Enter notes here"
      />

      <div className="text-start mb-3 flex-grow">
        <Button
          disabled={newNote === ""}
          onClick={validateAndAdd}
          variant="primary mt-2">
          Save Notes
        </Button>
      </div>
    </>
  );
}

export default ServiceNotes;
