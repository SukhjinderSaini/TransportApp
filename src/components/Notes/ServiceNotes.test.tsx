import React from "react";
import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";

import ServiceNotes from "./ServiceNotes";
it("Notes renders correctly", () => {
  const tree = renderer
    .create(<ServiceNotes serviceName="Sydney Bus" />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("shows service notes text area correctly", () => {
  render(<ServiceNotes serviceName="Sydney Bus" />);
  expect(screen.getByRole("textbox")).toBeInTheDocument();
});

it("shows service add button correctly", () => {
  render(<ServiceNotes serviceName="Sydney Bus" />);
  expect(screen.getByRole("button")).toBeDisabled();
});
