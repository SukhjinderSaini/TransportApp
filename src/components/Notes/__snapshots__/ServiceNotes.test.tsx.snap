// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notes renders correctly 1`] = `
Array [
  <div
    className="list-group"
    onKeyDown={[Function]}
  />,
  <textarea
    aria-label="With textarea"
    className="mt-3 mb-3 form-control"
    onChange={[Function]}
    placeholder="Enter notes here"
    value=""
  />,
  <div
    className="text-start mb-3 flex-grow"
  >
    <button
      className="btn btn-primary mt-2"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      Save Notes
    </button>
  </div>,
]
`;
